### PR TITLE
smart amp: added support for static DSM library

### PR DESCRIPTION
**Pull request by ryans-lee**
Added support for static DSM library

Added Maxim DSM(Dyanmic Speaker Management) static library and its integration code.
- License information of the Maxim DSM library can be found in src/audio/smart_amp/dsm_lib/LICENSE.dsm
- Library usage and ISA information can be found in src/audio/smart_amp/dsm_lib/README.dsm
- Library do not have dependency with SOF version. This information was added to README.dsm.

Signed-off-by: Ryan Lee <ryans.lee@maximintegrated.com>

### CODE REVIEW

**Comment by ryans-lee** [comment]
This PR was originally discussed from another [PR#3302](https://github.com/thesofproject/sof/pull/3302) to different branch.
Now created new PR to the master.

File: src/audio/smart_amp/smart_amp_maxim_dsm.c [hunk 1]
```diff
@@ -18,11 +18,48 @@
...
 
 static int maxim_dsm_init(struct smart_amp_mod_struct_t *hspk, struct comp_dev *dev)
 {
-	/* Maxim Dynamic Speaker Manangement initialization */
+	int *circularbuffersize = hspk->circularbuffersize;
+	int *delayedsamples = hspk->delayedsamples;
+	struct dsm_api_init_ext_t initparam;
+	enum DSM_API_MESSAGE retcode;
+
+	comp_dbg(dev, BUILD_TIME);
```
Comment (lgirdwood) [inline src/audio/smart_amp/smart_amp_maxim_dsm.c:30]: Should this be removed ? It will make the build non reproducible.

File: src/audio/smart_amp/smart_amp_maxim_dsm.c [hunk 2]
```diff
@@ -173,20 +211,20 @@ static void maxim_dsm_fb_proc(struct smart_amp_mod_struct_t *hspk,
...
-			/* Buffer ordering for DSM : IVIV... -> VV... II...*/
-				v[x] = buf.buf16[4 * x + 1];
-				i[x] = buf.buf16[4 * x];
-				v[x + DSM_FRM_SZ] = buf.buf16[4 * x + 3];
-				i[x + DSM_FRM_SZ] = buf.buf16[4 * x + 2];
+			/* Buffer ordering for DSM : VIVI... -> VV... II...*/
+				v[x] = buf.buf16[4 * x];
+				i[x] = buf.buf16[4 * x + 1];
+				v[x + DSM_FRM_SZ] = buf.buf16[4 * x + 2];
+				i[x + DSM_FRM_SZ] = buf.buf16[4 * x + 3];
```
Comment (lgirdwood) [inline src/audio/smart_amp/smart_amp_maxim_dsm.c:218]: These look like separate fixes and should be in a different patch to license and binary.

**Review by lgirdwood — CHANGES_REQUESTED** [review]
@ryans-lee thanks, can you split this into more patches as it looks like a mix of fixes are included with the binary. I will have someone check the license to make sure its good here too.

File: src/audio/smart_amp/smart_amp_maxim_dsm.c (hunk 1 above)
Comment (ryans-lee) [reply src/audio/smart_amp/smart_amp_maxim_dsm.c:30]: Agreed. It should be removed.

File: src/audio/smart_amp/smart_amp_maxim_dsm.c (hunk 2 above)
Comment (ryans-lee) [reply src/audio/smart_amp/smart_amp_maxim_dsm.c:218]: OK. Shall split the patch. Thanks.

**Comment by ryans-lee** [comment]
@lgirdwood  I split the patch in three.  Added comments on each commit. Thanks.

**Comment by lgirdwood** [comment]
@ryans-lee can you remove the binary and licence parts and we can merge the code changes only. I'm still checking the licence.

**Comment by ryans-lee** [comment]
> @ryans-lee can you remove the binary and licence parts and we can merge the code changes only. I'm still checking the licence.

@lgirdwood  I can do it, but I concern about the compile error if the library is removed. What do you think?

**Comment by lgirdwood** [comment]
@ryans-lee I guess the cmake part that links to the binary library can also be removed and form a new PR with the binary and license. This will minimise any parts that depend on the license.

**Comment by ryans-lee** [comment]
> @ryans-lee I guess the cmake part that links to the binary library can also be removed and form a new PR with the binary and license. This will minimise any parts that depend on the license.

@lgirdwood APIs in the binary library are used for the integration and this will cause link error without the library. Another option is making DSM component disabled as default from Kconfig to avoid build error until library and license file is merged. Let me update PR to apply this modification.

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
Jenkins CI on ICL is known issue.
